test/conformance/ingress/websocket_test.go

```diff
@@ -0,0 +1,199 @@
+// +build e2e
+
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ingress
+
+import (
+	"fmt"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/gorilla/websocket"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"knative.dev/serving/pkg/apis/networking/v1alpha1"
+	"knative.dev/serving/test"
+)
+
+// TestWebsocket verifies that websockets may be used via a simple Ingress.
+func TestWebsocket(t *testing.T) {
+	t.Parallel()
+	clients := test.Setup(t)
+
+	const suffix = "- pong"
+	name, port, cancel := CreateWebsocketService(t, clients, suffix)
+	defer cancel()
+
+	domain := name + ".example.com"
+
+	// Create a simple Ingress over the Service.
+	_, dialCtx, cancel := CreateIngressReadyDialContext(t, clients, v1alpha1.IngressSpec{
+		Rules: []v1alpha1.IngressRule{{
+			Hosts:      []string{domain},
+			Visibility: v1alpha1.IngressVisibilityExternalIP,
+			HTTP: &v1alpha1.HTTPIngressRuleValue{
+				Paths: []v1alpha1.HTTPIngressPath{{
+					Splits: []v1alpha1.IngressBackendSplit{{
+						IngressBackend: v1alpha1.IngressBackend{
+							ServiceName:      name,
+							ServiceNamespace: test.ServingNamespace,
+							ServicePort:      intstr.FromInt(port),
+						},
+					}},
+				}},
+			},
+		}},
+	})
+	defer cancel()
+
+	dialer := websocket.Dialer{
+		NetDialContext:   dialCtx,
+		Proxy:            http.ProxyFromEnvironment,
+		HandshakeTimeout: 45 * time.Second,
+	}
+
+	u := url.URL{Scheme: "ws", Host: domain, Path: "/"}
+	conn, _, err := dialer.Dial(u.String(), http.Header{"Host": {domain}})
+	if err != nil {
+		t.Fatalf("Dial() = %v", err)
+	}
+	defer conn.Close()
+
+	for i := 0; i < 100; i++ {
+		message := fmt.Sprintf("ping - %d", rand.Intn(1000))
```

Comment:
```suggestion
		message := "ping - " + strconv.Itoa(rand.Intn(1000))
```

Reply:
🤔 I think I generally prefer the readability of the format string vs. multiple other things.